### ISSUE TITLE
BeforeEach/AfterEach BeforeAll/AfterAll hooks

### ISSUE THREAD

**Issue by bandersongit** [facebookexperimental/reason-native#171]
This is a feature that is present in the Jest/xUnit testing frameworks ( as setup/teardown). This is useful for mocking out set up/ensuring things like network/db connections are closed regardless of exceptions.

I think that something like encoding data this is setup in beforeEach/beforeAll into a type argument that gets fed to describe/test makes sense as it would strongly discourage the use of refs (which is what I think is natural in JS) and we don't have a "this" object like a junit/nunit with properties that can be set.

The ergonomics of this are difficult and I'll update this issue with some thoughts/am very open to whatever thoughts anyone else may have

**Comment by bandersongit** [comment]
@kyldvs and I came up with an implementation that I will be implementing soon.

High level overview

```reason
/* t will actually be t('beforeAll, 'afterAll, 'beforeEach, 'afterEach, 'all, 'each) and the lifecycle methods will be typed to ensure that they aren't called multiple times/in an order that doesn't make sense)
module TestLifeCycle: {
  type t;
  let beforeAll: (unit => 'a, t) => t;
  let afterAll: (unit => 'a, t) => t;
  let beforeEach: ('all => 'a, t) => t;
  let afterEach: ('each => 'a, 't) => t;
  let toTestSuite: t => {
    describe,
    describeOnly,
    describeSkip
  }
}
```

The "test environment" will be persisted through test blocks inside describes created this way via a new field on the record that contains expect.

E.g. usage will look like
```reason
open TestFramework;

let {describe} = testLifecycle
|> beforeAll(() => { foo: barFactory.create()  })
|> afterAll({foo} => foo.dispose())
|> toTestSuite;

describe("my test suite with beforeAll and afterAll", ({test}) => {
  test("some test", ({expect, env}) => {
    let myBar = env.foo;
    expect.string(myBar.description).toBe("look how the setup flows through");
  });
});

let {describe} = testLifecycle
|> beforeAll(() => { foo: barFactory.create()  })
|> beforeEach(({foo}) => foo.id)
|> afterEach(() => print_endline("I don't have to do anything here if I don't want to")
|> afterAll(({foo}) => foo.dispose());
|> toTestSuite;

describe("my test suite with beforeEach and afterEach", ({test}) => {
  test("some test", ({expect, env}) => {
    let myBar = env.foo;
    expect.int(myBar.id).toBe(42);
  });
});
```

**Comment by bandersongit** [comment]
This has been released as part of Rely 3.0.0, documentation can be found [here](https://reason-native.com/docs/rely/setup-teardown)
